File: services/python/shared/middleware/idempotency.py
"""
Idempotency Service for Ag-Fintech Platform
Provides idempotent operation handling using Redis
"""

import hashlib
import json
import time
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import Any, Optional
import redis


@dataclass
class IdempotencyResult:
    """Result of an idempotent operation"""
    key: str
    status: str  # "new", "processing", "completed", "failed"
    result: Optional[Any] = None
    error: Optional[str] = None
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)

    def to_dict(self) -> dict:
        return {
            "key": self.key,
            "status": self.status,
            "result": self.result,
            "error": self.error,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }

    @classmethod
    def from_dict(cls, data: dict) -> "IdempotencyResult":
        return cls(
            key=data["key"],
            status=data["status"],
            result=data.get("result"),
            error=data.get("error"),
            created_at=datetime.fromisoformat(data["created_at"]),
            updated_at=datetime.fromisoformat(data["updated_at"]),
        )


def generate_key(prefix: str, *identifiers) -> str:
    """Generate a deterministic idempotency key from business identifiers"""
    data = f"{prefix}:" + ":".join(str(id) for id in identifiers)
    hash_bytes = hashlib.sha256(data.encode()).digest()
    return hash_bytes[:16].hex()  # Use first 16 bytes for shorter key


def generate_transfer_id(farmer_id: int, entity_type: str, entity_id: int, sequence: int) -> int:
    """Generate a deterministic transfer ID for TigerBeetle"""
    key = f"transfer:{farmer_id}:{entity_type}:{entity_id}:{sequence}"
    hash_bytes = hashlib.sha256(key.encode()).digest()
    # Use first 8 bytes as uint64
    return int.from_bytes(hash_bytes[:8], byteorder='big')


def generate_account_id(farmer_id: int, account_type: int) -> int:
    """Generate a deterministic account ID for TigerBeetle"""
    return farmer_id * 10000 + account_type


class IdempotencyService:
    """Provides idempotent operation handling"""

    def __init__(self, redis_client: redis.Redis, ttl: timedelta = timedelta(hours=24)):
        self.redis = redis_client
        self.ttl = ttl

    def _get_key(self, key: str) -> str:
        return f"idempotency:{key}"

    def try_acquire(self, key: str) -> tuple[bool, Optional[IdempotencyResult]]:
        """
        Attempt to acquire an idempotency lock.
        Returns (is_new, existing_result)
        """
        idemp_key = self._get_key(key)

        # Try to get existing result
        existing = self.redis.get(idemp_key)
        if existing:
            result = IdempotencyResult.from_dict(json.loads(existing))
            return False, result

        # Key doesn't exist - try to acquire lock with SETNX
        now = datetime.utcnow()
        new_result = IdempotencyResult(
            key=key,
            status="processing",
            created_at=now,
            updated_at=now,
        )
        data = json.dumps(new_result.to_dict())

        # Use SETNX to atomically set if not exists
        success = self.redis.setnx(idemp_key, data)
        if success:
            self.redis.expire(idemp_key, int(self.ttl.total_seconds()))
            return True, None

        # Another process acquired the lock - get their result
        existing = self.redis.get(idemp_key)
        if existing:
            result = IdempotencyResult.from_dict(json.loads(existing))
            return False, result

        return False, None

    def complete(self, key: str, result: Any) -> None:
        """Mark an idempotent operation as completed with result"""
        idemp_key = self._get_key(key)

        completed_result = IdempotencyResult(
            key=key,
            status="completed",
            result=result,
            updated_at=datetime.utcnow(),
        )
        data = json.dumps(completed_result.to_dict())
        self.redis.setex(idemp_key, int(self.ttl.total_seconds()), data)

    def fail(self, key: str, error_msg: str) -> None:
        """Mark an idempotent operation as failed"""
        idemp_key = self._get_key(key)

        failed_result = IdempotencyResult(
            key=key,
            status="failed",
            error=error_msg,
            updated_at=datetime.utcnow(),
        )
        data = json.dumps(failed_result.to_dict())
        self.redis.setex(idemp_key, int(self.ttl.total_seconds()), data)

    def release(self, key: str) -> None:
        """Release an idempotency lock (for cleanup on error before completion)"""
        idemp_key = self._get_key(key)
        self.redis.delete(idemp_key)


class ProcessedEventsTracker:
    """Tracks processed Kafka/Fluvio events for exactly-once semantics"""

    def __init__(self, redis_client: redis.Redis, ttl: timedelta = timedelta(days=7)):
        self.redis = redis_client
        self.ttl = ttl

    def _get_key(self, event_id: str) -> str:
        return f"processed_event:{event_id}"

    def is_processed(self, event_id: str) -> bool:
        """Check if an event has already been processed"""
        key = self._get_key(event_id)
        return self.redis.exists(key) > 0

    def mark_processed(self, event_id: str) -> None:
        """Mark an event as processed"""
        key = self._get_key(event_id)
        self.redis.setex(key, int(self.ttl.total_seconds()), int(time.time()))


class DistributedLock:
    """Provides distributed locking for non-atomic operations"""

    def __init__(self, redis_client: redis.Redis, resource: str, ttl: timedelta):
        self.redis = redis_client
        self.key = f"lock:{resource}"
        self.value = f"{time.time_ns()}-{time.time_ns() % 1000000}"
        self.ttl = ttl

    def acquire(self) -> bool:
        """Attempt to acquire the lock"""
        return bool(self.redis.set(
            self.key,
            self.value,
            nx=True,
            px=int(self.ttl.total_seconds() * 1000)
        ))

    def release(self) -> bool:
        """Release the lock (only if we own it)"""
        script = """
        if redis.call("get", KEYS[1]) == ARGV[1] then
            return redis.call("del", KEYS[1])
        else
            return 0
        end
        """
        return bool(self.redis.eval(script, 1, self.key, self.value))

    def extend(self, ttl: timedelta) -> bool:
        """Extend the lock TTL (for long-running operations)"""
        script = """
        if redis.call("get", KEYS[1]) == ARGV[1] then
            return redis.call("pexpire", KEYS[1], ARGV[2])
        else
            return 0
        end
        """
        return bool(self.redis.eval(script, 1, self.key, self.value, int(ttl.total_seconds() * 1000)))

    def __enter__(self):
        if not self.acquire():
            raise RuntimeError(f"Failed to acquire lock: {self.key}")
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.release()
        return False

File: services/go/shared/middleware/idempotency.go
package middleware

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

// IdempotencyService provides idempotent operation handling
type IdempotencyService struct {
	redis *redis.Client
	ttl   time.Duration
}

// IdempotencyResult represents the result of an idempotent operation
type IdempotencyResult struct {
	Key       string      `json:"key"`
	Status    string      `json:"status"` // "new", "processing", "completed", "failed"
	Result    interface{} `json:"result,omitempty"`
	Error     string      `json:"error,omitempty"`
	CreatedAt time.Time   `json:"created_at"`
	UpdatedAt time.Time   `json:"updated_at"`
}

// NewIdempotencyService creates a new idempotency service
func NewIdempotencyService(redisClient *redis.Client, ttl time.Duration) *IdempotencyService {
	if ttl == 0 {
		ttl = 24 * time.Hour // Default 24 hour TTL
	}
	return &IdempotencyService{
		redis: redisClient,
		ttl:   ttl,
	}
}

// GenerateKey generates a deterministic idempotency key from business identifiers
func GenerateKey(prefix string, identifiers ...interface{}) string {
	data := fmt.Sprintf("%s:", prefix)
	for _, id := range identifiers {
		data += fmt.Sprintf("%v:", id)
	}
	hash := sha256.Sum256([]byte(data))
	return hex.EncodeToString(hash[:16]) // Use first 16 bytes for shorter key
}

// GenerateTransferID generates a deterministic transfer ID for TigerBeetle
// This ensures the same logical transfer always gets the same ID
func GenerateTransferID(farmerId int, entityType string, entityId int, sequence int) uint64 {
	key := fmt.Sprintf("transfer:%d:%s:%d:%d", farmerId, entityType, entityId, sequence)
	hash := sha256.Sum256([]byte(key))
	// Use first 8 bytes as uint64
	var id uint64
	for i := 0; i < 8; i++ {
		id = (id << 8) | uint64(hash[i])
	}
	return id
}

// GenerateAccountID generates a deterministic account ID for TigerBeetle
func GenerateAccountID(farmerId int, accountType int) uint64 {
	return uint64(farmerId)*10000 + uint64(accountType)
}

// TryAcquire attempts to acquire an idempotency lock
// Returns (isNew, existingResult, error)
func (s *IdempotencyService) TryAcquire(ctx context.Context, key string) (bool, *IdempotencyResult, error) {
	idempKey := fmt.Sprintf("idempotency:%s", key)

	// Try to get existing result
	existing, err := s.redis.Get(ctx, idempKey).Result()
	if err == nil {
		// Key exists - return existing result
		var result IdempotencyResult
		if err := json.Unmarshal([]byte(existing), &result); err != nil {
			return false, nil, fmt.Errorf("failed to unmarshal existing result: %w", err)
		}
		return false, &result, nil
	}
	if err != redis.Nil {
		return false, nil, fmt.Errorf("redis error: %w", err)
	}

	// Key doesn't exist - try to acquire lock with SETNX
	now := time.Now()
	newResult := IdempotencyResult{
		Key:       key,
		Status:    "processing",
		CreatedAt: now,
		UpdatedAt: now,
	}
	data, _ := json.Marshal(newResult)

	// Use SETNX to atomically set if not exists
	success, err := s.redis.SetNX(ctx, idempKey, data, s.ttl).Result()
	if err != nil {
		return false, nil, fmt.Errorf("failed to acquire lock: %w", err)
	}

	if !success {
		// Another process acquired the lock - get their result
		existing, err := s.redis.Get(ctx, idempKey).Result()
		if err != nil {
			return false, nil, fmt.Errorf("failed to get concurrent result: %w", err)
		}
		var result IdempotencyResult
		if err := json.Unmarshal([]byte(existing), &result); err != nil {
			return false, nil, fmt.Errorf("failed to unmarshal concurrent result: %w", err)
		}
		return false, &result, nil
	}

	return true, nil, nil
}

// Complete marks an idempotent operation as completed with result
func (s *IdempotencyService) Complete(ctx context.Context, key string, result interface{}) error {
	idempKey := fmt.Sprintf("idempotency:%s", key)

	completedResult := IdempotencyResult{
		Key:       key,
		Status:    "completed",
		Result:    result,
		UpdatedAt: time.Now(),
	}
	data, _ := json.Marshal(completedResult)

	return s.redis.Set(ctx, idempKey, data, s.ttl).Err()
}

// Fail marks an idempotent operation as failed
func (s *IdempotencyService) Fail(ctx context.Context, key string, errMsg string) error {
	idempKey := fmt.Sprintf("idempotency:%s", key)

	failedResult := IdempotencyResult{
		Key:       key,
		Status:    "failed",
		Error:     errMsg,
		UpdatedAt: time.Now(),
	}
	data, _ := json.Marshal(failedResult)

	return s.redis.Set(ctx, idempKey, data, s.ttl).Err()
}

// Release releases an idempotency lock (for cleanup on error before completion)
func (s *IdempotencyService) Release(ctx context.Context, key string) error {
	idempKey := fmt.Sprintf("idempotency:%s", key)
	return s.redis.Del(ctx, idempKey).Err()
}

// ProcessedEventsTracker tracks processed Kafka/Fluvio events for exactly-once semantics
type ProcessedEventsTracker struct {
	redis *redis.Client
	ttl   time.Duration
}

// NewProcessedEventsTracker creates a new event tracker
func NewProcessedEventsTracker(redisClient *redis.Client, ttl time.Duration) *ProcessedEventsTracker {
	if ttl == 0 {
		ttl = 7 * 24 * time.Hour // Default 7 day TTL for event deduplication
	}
	return &ProcessedEventsTracker{
		redis: redisClient,
		ttl:   ttl,
	}
}

// IsProcessed checks if an event has already been processed
func (t *ProcessedEventsTracker) IsProcessed(ctx context.Context, eventID string) (bool, error) {
	key := fmt.Sprintf("processed_event:%s", eventID)
	exists, err := t.redis.Exists(ctx, key).Result()
	if err != nil {
		return false, err
	}
	return exists > 0, nil
}

// MarkProcessed marks an event as processed
func (t *ProcessedEventsTracker) MarkProcessed(ctx context.Context, eventID string) error {
	key := fmt.Sprintf("processed_event:%s", eventID)
	return t.redis.Set(ctx, key, time.Now().Unix(), t.ttl).Err()
}

// DistributedLock provides distributed locking for non-atomic operations
type DistributedLock struct {
	redis  *redis.Client
	key    string
	value  string
	ttl    time.Duration
}

// NewDistributedLock creates a new distributed lock
func NewDistributedLock(redisClient *redis.Client, resource string, ttl time.Duration) *DistributedLock {
	return &DistributedLock{
		redis: redisClient,
		key:   fmt.Sprintf("lock:%s", resource),
		value: fmt.Sprintf("%d-%d", time.Now().UnixNano(), time.Now().UnixNano()%1000000),
		ttl:   ttl,
	}
}

// Acquire attempts to acquire the lock
func (l *DistributedLock) Acquire(ctx context.Context) (bool, error) {
	return l.redis.SetNX(ctx, l.key, l.value, l.ttl).Result()
}

// Release releases the lock (only if we own it)
func (l *DistributedLock) Release(ctx context.Context) error {
	// Use Lua script to atomically check and delete
	script := `
		if redis.call("get", KEYS[1]) == ARGV[1] then
			return redis.call("del", KEYS[1])
		else
			return 0
		end
	`
	_, err := l.redis.Eval(ctx, script, []string{l.key}, l.value).Result()
	return err
}

// Extend extends the lock TTL (for long-running operations)
func (l *DistributedLock) Extend(ctx context.Context, ttl time.Duration) error {
	script := `
		if redis.call("get", KEYS[1]) == ARGV[1] then
			return redis.call("pexpire", KEYS[1], ARGV[2])
		else
			return 0
		end
	`
	_, err := l.redis.Eval(ctx, script, []string{l.key}, l.value, ttl.Milliseconds()).Result()
	return err
}
